{
  "cells": [
    {
      "cell_type": "raw",
      "metadata": {},
      "source": [
        "---\n",
        "title: \"Introduction to F-Statistics\"\n",
        "author: \"Stephan Schiffels\"\n",
        "date: \"10/24/2023\"\n",
        "bibliography: references.bib\n",
        "toc: true\n",
        "reference-location: margin\n",
        "citation-location: margin\n",
        "\n",
        "---"
      ],
      "id": "89a23212"
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "## Introduction\n",
        "\n",
        "F-Statistics are one of the simplest summary-statistics of genotype data that help us gain insights into the relationships of individuals and groups.\n",
        "\n",
        "Some key references up front:\n",
        "\n",
        "1. Patterson, Nick, Priya Moorjani, Yontao Luo, Swapan Mallick, Nadin Rohland, Yiping Zhan, Teri Genschoreck, Teresa Webster, and David Reich. 2012. “Ancient Admixture in Human History.” Genetics 192 (3): 1065–93. [@Patterson2012]\n",
        "2. Peter, Benjamin M. 2016. “Admixture, Population Structure, and F-Statistics.” Genetics 202 (4): 1485–1501. [@Peter2016] \n",
        "\n",
        "## Outgroup-F3-Statistics\n",
        "\n",
        "Outgroup-F3-Statistics are perhaps one of the easiest to understand, so we'll start with them. F3-Statistics in general are defined as:\n",
        "\n",
        "$$F_3(A, B;C) = \\langle(c-a)(c-b)\\rangle$$\n",
        "\n",
        "which looks a bit unintuitive. Here, $A$ and $B$ are populations or individuals to be compared, and $C$ denotes an outgroup, and $a$ and $b$ are the non-reference allele frequencies of $A$ and $B$, respectively. The expression $\\langle\\cdot\\rangle$ denotes an average over all SNPs in the dataset.\n",
        "\n",
        "We can understand it best as a measure of genetic similarity between groups A and B, evaluated on SNPs where A and B carry a different allele than the outgroup C. Here is an example based on A, B and C all consisting of just a single haploid sample each:\n",
        "\n",
        "| SNP           | A | B | C | $(c-a)(c-b)$ |\n",
        "|---------------|---|---|---|--------------|\n",
        "| 1             | 1 | 1 | 0 | 1            |\n",
        "| 2             | 1 | 0 | 0 | 0            |\n",
        "| 3             | 0 | 1 | 0 | 0            |\n",
        "| 4             | 0 | 1 | 1 | 0            |\n",
        "| 5             | 0 | 0 | 1 | 1            |\n",
        "| 6             | 1 | 0 | 1 | 0            |\n",
        "| $F_3(A, B;C)$ |   |   |   | **0.33**     |\n",
        "\n",
        "Every SNP where A and B have the same allele, which in turn is different from the outgroup C contributes 1 to the statistic, all other positions contribute 0. So the higher this number, the more similar A and B are, and the more distant is the outgroup C from both A and B.\n",
        "\n",
        "## \"Linearity\" in A and B (but not C)\n",
        "What happens if $A$ and $B$ are populations, instead of single samples? Conveniently, this statistic is \"linear\" in A and B, which in this context means that the resulting statistic will just be an average over all combinations of individuals in A and B. The same is true for diploid samples A and B, in which case the result will just be the average similarity between all haplotypes between A and B. In mathematical terms:\n",
        "\n",
        "$$F_3(\\{A_1,A_2\\},B;C) = \\frac{1}{2}(F_3(A_1,B;C)+F_3(A_2,B;C))$$\n",
        "\n",
        "(and similarly for $B=\\{B_1,B_2\\}).\n",
        "\n",
        "Note that F_3 statistics are _not_ linear in $C$. This has rarely any consequences, because we typically use some fixed group for $C$ (see below), and only vary $A$ and $B$ in our analyses. But it is useful for the interpration to know that at least for A and B, group-based statistics are simply averages over individual-based ones.\n",
        "\n",
        "## How to use Outgroup-F3 - An example from [@Raghavan2014]\n",
        "\n",
        "Outgroup-F3 are most commonly used to characterize an individual or population of interest by its relationship to a set of _reference_ populations, for example to reveal geographic or cultural patterns of similarity.\n",
        "\n",
        "An example is this Figure from [@Raghavan2014]:\n",
        "\n",
        "![Figure 1c from [@Raghavan2014]](img/fstats/Raghavan2014_f3.jpg)\n",
        "\n",
        "which report $F_3(MA\\_1,X; \\text{Yoruba})$, where \"MA_1\" is an Upper Paleolithic individual from Siberia, approximately 24,000 years old, and Yoruba is a present-day group from Nigeria, which is here used as an outgroup.\n",
        "\n",
        "We can reproduce this statistic using `xerxes fstat`.\n",
        "\n",
        ":::{.callout-note}\n",
        "If you happen to have downloaded a copy of our [community archive](https://www.poseidon-adna.org/#/archive_overview?id=public-poseidon-archives) already, then just use the path to that archive in the following commands. Otherwise you download the entire archive via \n",
        "\n",
        "\n",
        "```{bash}\n",
        "trident fetch -d /somewhere/to/store/the/archive --downloadAll \n",
        "```\n",
        "\n",
        "\n",
        "or just the relevant packages for the examples in this chapter:\n",
        "trident fetch -d /somewhere/to/store/the/archive -f \"*2012_PattersonGenetics*,*2014_RaghavanNature*,*2014_LazaridisNature*,*2018_Lamnidis_Fennoscandia*\"\n",
        ":::\n",
        "\n",
        "You can "
      ],
      "id": "e37b8000"
    }
  ],
  "metadata": {
    "kernelspec": {
      "display_name": "Python 3",
      "language": "python",
      "name": "python3"
    }
  },
  "nbformat": 4,
  "nbformat_minor": 5
}