# Genotype and context data management with Poseidon

Poseidon is an open computational framework to enable standardised and FAIR (@Wilkinson2016) handling of genotypes with their highly relevant context information. It includes a well-specified data format, advanced software tools, and public, community-maintained archives to support the entire archaeogenetic research cycle, from data acquisition to management, analysis and publication.

Poseidon and all its components is well documented at <https://www.poseidon-adna.org>. In this tutorial we will give a brief overview and describe two specific workflows in the context of Poseidon.

## The components of Poseidon

![Overview of the Poseidon framework](img/poseidon/popgen_toolbox_schema.png){width=60%}

Poseidon is an entire ecosystem build on top of the data format specification of the Poseidon package.

### The Poseidon package

A [**Poseidon package**](https://www.poseidon-adna.org/#/standard) bundles [genotype data in EIGENSTRAT/PLINK format](https://www.poseidon-adna.org/#/genotype_data) with human- and machine-readable meta-data.

![The files in a Poseidon package](img/poseidon/popgen_toolbox_package_schema.png){width=60%}

This includes sample-wise context like spatio-temporal origin and genetic data quality in [the `.janno`-file](https://www.poseidon-adna.org/#/janno_details), literature in the `.bib`, and pointers to sequencing data in [the `.ssf` file](https://www.poseidon-adna.org/#/ssf_details).

`.janno` and `.ssf` have predefined variables, but can store arbitrary additional information.

### The software tools

[**trident**](https://www.poseidon-adna.org/#/trident) is a command line application to create, download, inspect and merge Poseidon packages â€“ and therefore the central tool of the Poseidon framework. The `init` subcommand creates new packages from genotype data, `fetch` downloads them from the public archives through the Web-API, and `forge` merges and subsets them as specified. `list` gives an overview over entities in a set of packages and `validate` confirms their structural integrity.

[**xerxes**](https://www.poseidon-adna.org/#/xerxes) is derived from trident and allows to directly perform various basic and experimental genomic data analyses on Poseidon packages. It implements allele sharing statistics ($F_2$, $F_3$, $F_4$, $F_{ST}$) with a flexible permutation interface.

[**janno**](https://www.poseidon-adna.org/#/janno_r_package) is an R package to simplify reading .janno files into R and the popular tidyverse ecosystem (@Wickham2019). It provides an S3 class `janno` that inherits from `tibble`.

[**qjanno**](https://www.poseidon-adna.org/#/qjanno) is another command line tool to perform SQL queries on `.janno` files. On startup it creates a database in memory and reads .janno files into it. It then sends any user-provided SQL query to the database server and forwards its output.

### The public archives

The Poseidon community maintains [**public archives**](https://www.poseidon-adna.org/#/archive_overview) for Poseidon packages to establish a central, open access point for published data in Poseidon format.

- The **Community Archive**: Per-paper packages with the genotype data published in the respective papers, partially pre-populated from the AADR. The community archive focusses on author submissions.
- The **AADR Archive**: Complete and structurally unified AADR releases repackaged in the Poseidon package format.
- The **Minotaur Archive**: Per-paper packages with genotype data reprocessed by the Minotaur workflow (see below).

The data is versioned with Git and hosted on GitHub for easy co-editing and automatic structural validation. It can be accessed through a [**Web-API**](https://www.poseidon-adna.org/#/web_api) with various endpoints at *server.poseidon-adna.org*.

This API enables an [**Archive explorer**](https://www.poseidon-adna.org/#/archive_explorer) on the Poseidon website.

### The Minotaur workflow

The [**Minotaur workflow**](https://www.poseidon-adna.org/#/minotaur) is a semi-automatic workflow to reproducibly process published sequencing data from the [International Nucleotide Sequence Database Collaboration](Ihttps://www.insdc.org) archives into Poseidon packages.

Community members can request new packages by submitting a build recipe as a Pull Request against a dedicated GitHub repository. This recipe is created from a Sequencing Source File (`.ssf`), describing the sequencing data for the package and where it can be downloaded.

Using the recipe, the sequencing data gets processed through nf-core/eager (@FellowsYates2021) on computational infrastructure of MPI-EVA, using a standardised, yet flexible, set of parameters.

The generated genotypes, together with descriptive statistics of the sequencing data (Endogenous, Damage, Nr_SNPs, Contamination), are compiled into a Poseidon package, and made available to users in the Minotaur Archive.

## Forging a dataset with `trident`

## Contributing to the community archive

![Poseidon needs you for public data maintenance](img/poseidon/recruiting.jpg){width=40%}

To maintain the public data archives, specifically the community archive and the minotaur archive, Poseidon depends on work donations by an interested community. Many practitioners of archaeogenetics both produce genotype data from archaological contexts and require the reference data from other publications, provided in public archives, to contextualize it.

If authors themselves provide high-quality, easily accessible versions of their data they gain at least three important advantages:

- Their work will be easily findable and potentially cited more often.
- They have primacy over how exactly their data is communicated. That is, which genotypes, dates or group names they consider correct.
- Their results for derived, genotype based analyses (PCA, F-Statistics, etc.) can be reproduced exactly.

And the whole community wins, because sharing the tedious data preparation tasks empowers all researchers to achieve more in shorter time.

***

This tutorial explains the main cornerstones of a workflow to add a new Poseidon package to the community archive after publishing the corresponding dataset. The process is documented in detail in a [**Submission guide**](https://www.poseidon-adna.org/#/archive_submission_guide) on the website.

1. Make yourself familiar with a number of core technologies. This is less challenging than it sounds, because: Many of you already know a lot about them, superficial knowledge is sufficient and knowing them is useful for anybody who works with archaeogenetic data.

    - Editing text files.
    - 
    - Creating and validating Poseidon packages with the `trident` tool.
    - Free and open source distributed version control with [Git](https://git-scm.com).
    - Collaborative working on Git projects with [GitHub](https://github.com).
    - Handling large files in Git using [Git LFS](https://git-lfs.com).

2. ...
